independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ContextDataMap.java

```diff
@@ -0,0 +1,163 @@
+package io.quarkus.arc.impl;
+
+import io.quarkus.arc.ArcInvocationContext;
+import java.lang.annotation.Annotation;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import javax.interceptor.InvocationContext;
+
+/**
+ * This Map implementation is specifically optimised to implement
+ * {@link InvocationContext#getContextData()}.
+ * We try to leverage the fact that in most use cases, while there is a high
+ * chance for integration points to need to read the content of this map,
+ * write operations are less likely. So we only allocate an actual underlying
+ * HashMap on a write operation.
+ * In addition, an {@link ArcInvocationContext} always allows to return the set of
+ * interceptor bindings via {@link ArcInvocationContext#KEY_INTERCEPTOR_BINDINGS};
+ * this is implemented while avoiding a write operation on the underlying Map.
+ * While one goal is efficiency, another goal is safety: while the CDI specification
+ * allows any interceptor to make changes to this Map, we specifically disallow
+ * 3rd party code to make changes to the interceptor bindings of the platform.
+ */
+final class ContextDataMap implements Map<String, Object> {
+
+    private final Set<Annotation> interceptorBindings;
+    private HashMap<String, Object> delegate; //important to lazily initialize this
+
+    ContextDataMap(Set<Annotation> interceptorBindings) {
+        this.interceptorBindings = Objects.requireNonNull(interceptorBindings);
+    }
+
+    private Map<String, Object> getDelegateForRead() {
+        if (delegate == null) {
+            return Collections.EMPTY_MAP;
+        } else {
+            return delegate;
+        }
+    }
+
+    private Map<String, Object> getDelegateForWrite(final int sizeHint) {
+        if (delegate == null) {
+            this.delegate = new HashMap<>(sizeHint, 1.0f);
+        }
+        return delegate;
+    }
+
+    // ** Implement methods from the Map interface **
+
+    @Override
+    public void clear() {
+        throw new UnsupportedOperationException("Not allowed to clear the context data map");
+    }
+
+    @Override
+    public boolean containsKey(Object key) {
+        if (ArcInvocationContext.KEY_INTERCEPTOR_BINDINGS.equals(key)) {
+            return true;
+        }
+        return getDelegateForRead().containsKey(key);
+    }
+
+    @Override
+    public boolean containsValue(Object value) {
+        if (interceptorBindings.equals(value)) {
+            return true;
+        }
+        return getDelegateForRead().containsValue(value);
+    }
+
+    @Override
+    public Set<Map.Entry<String, Object>> entrySet() {
+        final AbstractMap.SimpleImmutableEntry<String, Object> firstEntry = new AbstractMap.SimpleImmutableEntry<>(
```

Comment:
This can be lazily allocated (and reused) as well, given that users cannot modify it?

Reply:
yes it could - but I don't think it's worth it: hopefully nobody is using `entrySet` heavily on this map.